### PR TITLE
attempt to improve performance

### PR DESCRIPTION
**Pull request by xobotyi**
- Fully reworked code, it's faster (2-2.5 times)
- Had to increase a complexity to 17;
- `editorconfig-tools` replaced with `eclint` due to inability to run tests on modern nodejs environments (tested on 11 and 12)

### CODE REVIEW

File: .istanbul.yml [hunk 1]
```diff
@@ -1,47 +1,47 @@
...
-    compact: true
-    preserve-comments: false
-    complete-copy: false
-    save-baseline: false
-    baseline-file: ./coverage/coverage-baseline.raw.json
-    include-all-sources: false
-    include-pid: false
-    es-modules: false
-    auto-wrap: false
+	root: .
```
Comment (ljharb) [inline .istanbul.yml]: Please revert unrelated formatting changes.

File: Makefile [hunk 2]
```diff
@@ -48,14 +48,14 @@ _ensure-clean:
...
-	   semver "$$new_ver" >/dev/null || { echo 'Invalid version number specified: $(TAG) - must be major.minor.patch' >&2; exit 2; }; \
-	   semver -r "> $$old_ver" "$$new_ver" >/dev/null || { echo 'Invalid version number specified: $(TAG) - must be HIGHER than current one.' >&2; exit 2; } \
-	 else \
-	   new_ver=`semver -i "$$new_ver" "$$old_ver"` || { echo 'Invalid version-increment specifier: $(TAG)' >&2; exit 2; } \
-	 fi; \
-	 printf "=== Bumping version **$$old_ver** to **$$new_ver** before committing and tagging:\n=== TYPE 'proceed' TO PROCEED, anything else to abort: " && read response && [ "$$response" = 'proceed' ] || { echo 'Aborted.' >&2; exit 2; };  \
-	 replace "$$old_ver" "$$new_ver" -- $(VERSIONED_FILES) && \
-	 git commit -m "v$$new_ver" $(VERSIONED_FILES) CHANGELOG.md && \
-	 git tag -a -m "v$$new_ver" "v$$new_ver"
+		new_ver=`echo "$(TAG)" | sed 's/^v//'`; new_ver=$${new_ver:-patch}; \
```
Comment (ljharb) [inline Makefile]: here as well.

File: index.js [hunk 3]
```diff
@@ -1,37 +1,28 @@
...
+	if (valueType === 'function') {
+		if (!value.prototype) { return true; } // arrow functions
+	} else if (valueType !== 'object') { return false; } // any else non function-ish types
 
-var tryFunctionObject = function tryFunctionToStr(value) {
 	try {
-		if (isES6ClassFn(value)) { return false; }
-		fnToStr.call(value);
-		return true;
+		var functionString = fnToStr.call(value); // most of non-functions will fail here
```
Comment (ljharb) [inline index.js]: by inlining the try/catch, this actually makes performance much much worse on everything that's not latest v8.

File: package.json [hunk 4]
```diff
@@ -31,7 +31,7 @@
 		"istanbul:merge": "istanbul-merge --out coverage/coverage.raw.json coverage-harmony/coverage.raw.json coverage-std/coverage.raw.json && istanbul report html",
 		"istanbul:harmony": "node --harmony ./node_modules/istanbul/lib/cli.js cover test.js --dir coverage-harmony",
 		"istanbul:std": "istanbul cover test.js --report html --dir coverage-std",
-		"prelint": "editorconfig-tools check *",
+		"prelint": "eclint check *",
```
Comment (ljharb) [inline package.json]: this is a good change, thanks - i'll cherrypick this in separately.

**Review by ljharb — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (xobotyi) [reply Makefile]: Just to be clear, what for? Simply running `npm run test` will fail due to eclint error and tests wont go further.

File: index.js (hunk 3 above)
Comment (xobotyi) [reply index.js]: Well, this is not a hard one too.
I'll fix it in a moment;

Comment (ljharb) [reply index.js]: what i mean is, the existing library in master afaik is the fastest it can possibly get, taking into account all supported browsers.

Comment (xobotyi) [reply index.js]: Sorry if bother, but why not as in my latest aproach?

Because for modern browsers (that has `toStringTag`) it calls `fnToStr.call(value);` twice: inside `isES6ClassFunction` and one more time inside `tryFunctionToStr`.  
Actually this is main reason why it is twice slower than it could be in modern browsers.

File: index.js [hunk 5]
```diff
@@ -1,37 +1,35 @@
...
-	if (isES6ClassFn(value)) { return false; }
-	var strClass = toStr.call(value);
-	return strClass === fnClass || strClass === genClass;
+
+	var tmp = typeof value;
+	if (tmp === 'function') {
+		if (!value.prototype) { return true; }
+	} else if (tmp !== 'object') { return false; }
+
+	if ((tmp = getFnStr(value)) === null) {
```
Comment (ljharb) [inline index.js]: please rewrite this without using assignment in expression position

Comment (xobotyi) [reply index.js]: done

File: index.js [hunk 6]
```diff
@@ -1,37 +1,36 @@
...
 	if (!value) { return false; }
-	if (typeof value !== 'function' && typeof value !== 'object') { return false; }
-	if (typeof value === 'function' && !value.prototype) { return true; }
-	if (hasToStringTag) { return tryFunctionObject(value); }
-	if (isES6ClassFn(value)) { return false; }
-	var strClass = toStr.call(value);
-	return strClass === fnClass || strClass === genClass;
+
+	var tmp = typeof value;
+	if (tmp === 'function') {
```
Comment (ljharb) [inline index.js]: storing the typeof value actually defeats engine optimizations in many engines (at least some older ones). it's better to repeat the `typeof value ===` every time.

Comment (xobotyi) [reply index.js]: Huh, didnt know.

File: Makefile (hunk 2 above)
Comment (xobotyi) [reply Makefile]: @ljharb As i said: https://travis-ci.org/ljharb/is-callable/jobs/551967918
Build failed cause makefile uses both, spaces and tabs

File: index.js (hunk 3 above)
Comment (ljharb) [reply index.js]: I agree that there's probably an improvement to be made in terms of calling Function toString, but that wouldn't be this large a rewrite.

Comment (xobotyi) [reply index.js]: Well initially i've just written it from scratch, to see if i can squeeze some performance without seing you code, thats why i has so much differences;

About main function changes:
- `lines 21-23`: little economy in strings comparisson, function check made only once instead of original twice compares.
- `lines 25-31`: the only way i see to merge two function calls into one.
- `lines 33-34`: no need to create one more variable if we can reuse existing, so why dont?

File: Makefile (hunk 2 above)
Comment (ljharb) [reply Makefile]: "everyone" most certainly doesn't :-) last time someone checked github, it was about 70% spaces, 30% tabs. Separately, tabs are more flexible and more accessible than spaces.

**Comment by ljharb** [comment]
Do you have benchmarks that justify these changes and can prove that they're faster?

**Comment by xobotyi** [comment]
@ljharb yep.
```javascript
'use strict';

/* eslint-disable */

var Benchmark = require('benchmark');
var local = require('../');
var npm = require('is-callable');

var suite = new Benchmark.Suite();

var regularFn = function () {};
var generator = function* () {};
var asyncFn = async function () {};
var scalar = 123;
var array = [1, 2, 3];
var regex = /tests/i;
var cls = class Foo {};

var return3 = function () { return 3; };
var fakeFunction = {
	toString: function () { return String(return3); },
	valueOf: return3
};

suite.add('local [function]', function () {
	local(regularFn);
});
suite.add('npm [function]', function () {
	npm(regularFn);
});

suite.add('local [fake function]', function () {
	local(fakeFunction);
});
suite.add('npm [fake function]', function () {
	npm(fakeFunction);
});

suite.add('local [async fn]', function () {
	local(asyncFn);
});
suite.add('npm [async fn]', function () {
	npm(asyncFn);
});

suite.add('local [generator]', function () {
	local(generator);
});
suite.add('npm [generator]', function () {
	npm(generator);
});

suite.add('local [class]', function () {
	local(cls);
});
suite.add('npm [class]', function () {
	npm(cls);
});

suite.add('local [scalar]', function () {
	local(scalar);
});
suite.add('npm [scalar]', function () {
	npm(scalar);
});

suite.add('local [array]', function () {
	local(array);
});
suite.add('npm [array]', function () {
	npm(array);
});

suite.add('local [regex]', function () {
	local(regex);
});
suite.add('npm [regex]', function () {
	npm(regex);
});

suite
	.on('cycle', function (event) { console.log(String(event.target)); })
	.run();
```

`local` is mine version `npm` is current npm version;

Results:
```
local [function] x 7,198,160 ops/sec ±0.33% (96 runs sampled)
npm [function] x 3,983,487 ops/sec ±0.44% (89 runs sampled)

local [fake function] x 110,051 ops/sec ±0.63% (86 runs sampled)
npm [fake function] x 48,741 ops/sec ±1.04% (78 runs sampled)

local [async fn] x 117,547,996 ops/sec ±0.59% (94 runs sampled)
npm [async fn] x 117,422,458 ops/sec ±0.25% (94 runs sampled)

local [generator] x 7,256,687 ops/sec ±0.10% (97 runs sampled)
npm [generator] x 3,956,934 ops/sec ±0.60% (96 runs sampled)

local [class] x 7,089,542 ops/sec ±1.19% (92 runs sampled)
npm [class] x 7,762,705 ops/sec ±0.22% (96 runs sampled)

local [scalar] x 918,646,647 ops/sec ±0.04% (95 runs sampled)
npm [scalar] x 917,248,841 ops/sec ±0.15% (94 runs sampled)

local [array] x 108,959 ops/sec ±0.37% (85 runs sampled)
npm [array] x 48,274 ops/sec ±0.83% (86 runs sampled)

local [regex] x 108,191 ops/sec ±0.37% (86 runs sampled)
npm [regex] x 48,760 ops/sec ±0.79% (86 runs sampled)
```

**Comment by xobotyi** [comment]
if to mix all the samples into single cycle call it will give this results:
```
local x 36,770 ops/sec ±0.98% (68 runs sampled)
npm x 14,346 ops/sec ±0.30% (55 runs sampled)
```
benched on node 11

**Comment by ljharb** [comment]
Based on those results, assuming higher is better, the ones that are meaningfully faster locally are `function`, `fake function`, `generator`, `scalar`, `array`, and `regex`; `class` is meaningfully slower.

Overall that still seems like an improvement for the most common cases. I'll have to dig into this more.
